### PR TITLE
Joint hypothesis tests

### PR DESCRIPTION
**Pull request by vincentarelbundock**
+ [x] print DF
+ [x] wildcards (does `data.table::patterns` help?)
+ [ ] linear combinations on the LHS
+ [x] RHS values (vector + examples in docs)
+ [x] Test with a single value

### CODE REVIEW

**Comment by vincentarelbundock** [comment]
No Bayés

**Comment by snhansen** [comment]
This is really nice, Vincent! I played around with it a bit and a couple of things came to mind.

A couple of feature requests:

* Be able to refer to all dummy variables induced by a term in the regression model specification. So for instance, if I write `joint = c("as.factor(cyl)")`, the function translates this to `joint = c("as.factor(cyl)6", "as.factor(cyl)8")` and if I write `joint = c("as.factor(cyl):hp")`, the function translates this to `joint = c("as.factor(cyl)6:hp", "as.factor(cyl)8:hp")`. As a reference, Stata's `test` function has this functionality and it's really nice not having to write out all terms (or write some code to do it) when a factor has many categories.

* Be able to specify the right hand side to something other than zero, e.g. `hypotheses(lmfit, joint = c("cyl6:hp = 3", "cyl8:hp = 2"), joint_test = "f")`.

* Be able to specify linear combinations on the left hand side, e.g. `hypotheses(lmfit, joint = c("cyl6+cyl6:hp", "cyl8+cyl8:hp"), joint_test = "f")`.


Other comments:

* When only one parameter is given it throws an error. It should probably just pass and provide the test statistics although it really isn't a joint test in this case.

* Might as well provide the degrees of freedom for both the F- and Chisq-test.

File: R/get_term_labels.R [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  if (is.data.frame(x)) {
+    if ("term" %in% names(x) && length(unique(x$term)) == nrow(x)) {
+      return(unique(x$term))
+    } else if (any(grepl("^contrast", names(x)))) {
+      tmp <- grep("^term$|^contrast", names(x))
+      out <- x[, tmp, drop = FALSE]
+      if (length(unique(out[["term"]])) == 1) {
+        out[["term"]] <- NULL
+      }
+      out <- apply(out, 1, paste, collapse = " ")
```
Comment (etiennebacher) [inline R/get_term_labels.R:11]: Just dropping by to say that this is faster and more memory efficient:
```r
out <- do.call(paste, c(out, sep = " "))
```

Comment (vincentarelbundock) [reply R/get_term_labels.R:11]: Old habits die hard. I now remember you teaching me this trick already. :)

File: R/get_term_labels.R [hunk 2]
```diff
@@ -0,0 +1,26 @@
+get_term_labels <- function(x, idx = NULL) {
+  if (is.data.frame(x)) {
+    if ("term" %in% names(x) && length(unique(x$term)) == nrow(x)) {
+      return(unique(x$term))
+    } else if (any(grepl("^contrast", names(x)))) {
```
[the remark this answers is not in the document]
Comment (vincentarelbundock) [reply R/get_term_labels.R:5]: But what if I like showing off my regex skills?

File: R/get_term_labels.R (hunk 1 above)
Comment (etiennebacher) [reply R/get_term_labels.R]: Yeah, I now have an alarm in my head for this one 😄

File: R/get_term_labels.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (strengejacke) [reply R/get_term_labels.R:5]: Then you should choose more impressive examples :-P

[the remark this answers is not in the document]
Comment (vincentarelbundock) [reply R/get_term_labels.R:5]: ok, that made me laugh :)

**Comment by vincentarelbundock** [comment]
@snhansen Thanks a lot for the suggestions. That's super useful.

All of the things are now implemented (ex: selection by regex), except for the linear combinations on the left hand side. This would be very hard to implement in the current code base, so it's out of scope for now.

Thanks again for the suggestion!

I expect that this will be released to CRAN in the next month, but it's available on Github now. You can see examples in the vignette:

https://vincentarelbundock.github.io/marginaleffects/dev/articles/hypothesis.html#joint-hyotheses-tests

**Comment by snhansen** [comment]
@vincentarelbundock: Looks good. I just played around with it a bit and it seems to work as intended. Very nice! I'll be using this in my teaching going forward.

Please do keep the linear combinations on the left hand side in mind for the future if the code base at some point permits it. This is something that comes up occasionally for me, so would be a nice feature to have.

**Comment by vincentarelbundock** [comment]
@snhansen, I just pushed a small commit to allow joint tests on linear combinations. The syntax is obviously not super convenient, but at least it’s now *possible*.

Example:

``` r
packageVersion("marginaleffects")
# [1] '0.11.2.9006'

library(marginaleffects)

# fit model
mod <- lm(mpg ~ factor(carb), mtcars)

# hypothesis matrix for linear combinations
H <- matrix(0, nrow = length(coef(mod)), ncol = 2)
H[2:3, 1] <- H[4:6, 2] <- 1

# test individual linear combinations
hyp <- hypotheses(mod, hypothesis = H)
hyp
# 
#    Term Estimate Std. Error     z Pr(>|z|) 2.5 % 97.5 %
#  custom    -12.0       4.92 -2.44  0.01477 -21.6  -2.35
#  custom    -25.5       9.03 -2.83  0.00466 -43.2  -7.85
# 
# Columns: term, estimate, std.error, statistic, p.value, conf.low, conf.high

# test joint hypotheses
hypotheses(hyp, joint = TRUE, hypothesis = c(-10, -20))
# 
# 
# Joint hypothesis test:
#  b1 = -10
#  b2 = -20
#  
#      F Pr(>|F|) Df 1 Df 2
#  0.197    0.822    2   30
# 
# Columns: statistic, p.value, df1, df2
```

**Comment by snhansen** [comment]
@vincentarelbundock: That's nice. Sure, the syntax could be more user-friendly, but it seems to get the job done, so I'll take it :-)

I'm not able to update to `0.11.2.9002`, however. If I do `remotes::install_github("vincentarelbundock/marginaleffects@joint")` then I'm getting `0.11.2.9002` which doesn't have this commit. Am I doing something wrong?

**Comment by strengejacke** [comment]
You can install 0.12.0 from CRAN.

**Comment by snhansen** [comment]
@strengejacke: Ah, didn't know it was already available on CRAN. Thanks.

**Comment by snhansen** [comment]
@vincentarelbundock: The denominator degrees of freedom aren't correct using the approach above. Here's a simple example where we get two different results for the same test:

```
library(marginaleffects)

lmfit <- lm(mpg~as.factor(cyl)*hp, data=mtcars)

H <- matrix(0, nrow = length(coef(lmfit)), ncol = 2)
H[5, 1] <- H[6, 2] <- 1
hyp <- hypotheses(lmfit, hypothesis = H)
hypotheses(hyp, joint = TRUE)
hypotheses(lmfit, joint = 5:6, joint_test = "f")
```

**Comment by vincentarelbundock** [comment]
Thanks for the report @snhansen I opened a separate issue with your example. Will look into it when I find time (probably not in the very very short term, unfortunately).

**Pull request merged**